xilem_core/src/state.rs

```diff
@@ -0,0 +1,219 @@
+///
+/// # Examples
+///
+/// A simple component can edit a single value when an event happens:
+///
+/// ```rust
+/// # use xilem_core::docs::{DocsView as WidgetView, stateless_component};
+/// # use xilem_core::map_state;
+/// use xilem_core::Edit;
+/// fn button() -> impl WidgetView<Edit<f64>> {
+///     // ...
+/// #   map_state(stateless_component(), |_, ()| ())
+/// }
+/// ```
+///
+/// Most components are likely to accept multiple parameters.
+/// Read-only parameters can use [`Read`].
+///
+/// ```rust
+/// # use xilem_core::docs::{DocsView as WidgetView, stateless_component};
+/// # use xilem_core::map_state;
+/// use core::ops::Range;
+/// use xilem_core::{Read, Edit};
+/// fn slider() -> impl WidgetView<(Edit<f64>, Read<Range<f64>>)> {
+///     // ...
+/// #   map_state(stateless_component(), |(_result, _range), ()| ())
+/// }
+/// ```
+// TODO: Maybe the name `Mut` here would make sense? I don't hate edit, though.
+// TODO: This forces T to be 'static, even though that isn't actually needed.
+#[doc(alias = "mutate")]
+#[doc(alias = "write")]
+pub type Edit<T> = &'static mut T;
+
+// TODO: These docs could probably do with a clarifying pass.
+/// When used in the first "state" parameter to `View`, this type indicates that
+/// the view value reads a value of type `T` when performing reconciliation and/or event handling.
+///
+/// This type implements [`ViewArgument`], and so any view operations for views which
+/// use this must be provided with a shared reference to `T` (i.e. `&T`).
+/// See [`Edit`] for the equivalent exclusive version of this.
+///
+/// # Examples
+///
+/// A simple component can display a single result:
+///
+/// ```rust
+/// # use xilem_core::docs::{DocsView as WidgetView, stateless_component};
+/// # use xilem_core::map_state;
+/// use xilem_core::Read;
+/// fn display_result() -> impl WidgetView<Read<f64>> {
+///     // ...
+/// #   map_state(stateless_component(), |_, ()| ())
+/// }
+/// ```
+///
+/// A component accepting only read access to a single value might not be that useful.
+/// Most components are likely to accept at least one [editable](Edit) parameter.
+/// This can be achieved using a tuple:
+///
+/// ```rust
+/// # use xilem_core::docs::{DocsView as WidgetView, stateless_component};
+/// # use xilem_core::map_state;
+/// use core::ops::Range;
+/// use xilem_core::{Read, Edit};
+/// fn slider() -> impl WidgetView<(Edit<f64>, Read<Range<f64>>)> {
+///     // ...
+/// #   map_state(stateless_component(), |(_result, _range), ()| ())
+/// }
+/// ```
+pub type Read<T> = &'static T;
+
+/// An alias to the "reference" form of a [`ViewArgument`].
+///
+/// This is used in `View` functions (instead of the expanded form) to decrease
+/// the amount of noise produced.
+// TODO: The name of this definitely needs rethinking.
+pub type Arg<'a, T> = <T as ViewArgument>::Params<'a>;
+
+// TODO: Do these need to be 'static?
```

Comment:
I think so? Otherwise `impl View` is not static is it (I think we had a discussion about this a while back)?

Neat trick by "abusing" a GAT with a `&'static` to avoid new type wrappers.